### PR TITLE
[api] Annotate as public / move ray-core APIs to _private and add enforcement rule

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?
 Enable checking of the `ray` core module, excluding serve, workflows, and tune, in `./ci/lint/check_api_annotations.py`. This required moving many files to `ray._private` and associated fixes.

### CODE REVIEW

**Review by richardliaw — APPROVED** [review]
stamp didnt read

File: ci/lint/check_api_annotations.py
Comment (pcmoritz) [inline ci/lint/check_api_annotations.py:33]: worth logging?

File: ci/lint/check_api_annotations.py
Comment (pcmoritz) [inline ci/lint/check_api_annotations.py:104]: ray.workflow?

File: python/ray/cloudpickle/cloudpickle_fast.py
Comment (pcmoritz) [inline python/ray/cloudpickle/cloudpickle_fast.py:25]: You probably want to leave this file untouched since it is copied from upstream, right?

**Review by pcmoritz — APPROVED** [review]
The cloudpickle changes should probably be reverted before merging.

**Comment by ericl** [comment]
Reverted the cloudpickle changes in favor of excluding it from the checker script for now.

**Comment by pcmoritz** [comment]
Thanks sounds great! Looks good to me, we can either try to get the codeowners or you ping me when the tests pass and we force-merge it.

**Review by clarkzinzow — APPROVED** [review]
Approving for data changes.

**Review by wuisawesome — APPROVED** [review]

**Comment by wuisawesome** [comment]
@DmitriGekhtman we should probably consider promoting `autoscaler.sdk.request_resources` to stable after this gets merged.

**Comment by DmitriGekhtman** [comment]
https://github.com/ray-project/ray/issues/22815

**Review by rkooo567 — APPROVED** [review]
We should also run nightly tests btw

**Review by gjoliver — APPROVED** [review]
RLlib is using ray._private.worker._mode() to detect whether we are in local or worker mode.
will there be a replacement API if this is explicitly marked as _private?

**Review by simon-mo — APPROVED** [review]
stamp

**Comment by ericl** [comment]
> RLlib is using ray._private.worker._mode() to detect whether we are in local or worker mode.

Given we're deprecating local mode, I think this may be something we can remove in the future.

**Pull request merged**
